sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala

```diff
@@ -122,6 +122,19 @@ object JdbcUtils extends Logging {
     }
   }
 
+  /**
+   * Runs a custom query against a table from the JDBC database.
+   */
+  def runQuery(conn: Connection, queryString: String, options: JDBCOptions): Unit = {
+    val statement = conn.prepareStatement(queryString)
```

Comment:
@moomindani Are we preventing user's to pass in "select queries" as part of pre or post actions ? If they pass in a query, then would we get an result set when we call "stme.execute()" that we need to free ? 

Another question is , what happens when a post action fails ? We don't have a notion of a transaction here, right ? In that case we probably need to document our behavior ?

Reply:
No, we do not prevent "select queries". Users can run any of "select", "insert", "update", "delete", etc.
Right, your concern is correct. I added implementation to close ResultSet for "select queries".

In addition, I added transaction and error handling logic. I also added unit tests for exceptions.